### PR TITLE
[WIP] Require detailed iterator types

### PR DESCRIPTION
**Pull request by thewilkybarkid**
Will be a follow-up to #32 to require annotations defining iterator types. Very much a WIP!

(https://github.com/slevomat/coding-standard/pull/558 was rejected.)

### CODE REVIEW

File: tests/cases/trait/method-defaults [hunk 1]
```diff
@@ -11,11 +11,11 @@ namespace Vendor;
 
 trait Foo
 {
-    public function bar($baz = [], $qux) : void
+    public function bar(string $baz = '', string $qux) : void
```
Comment (stephenwf) [inline tests/cases/trait/method-defaults]: Was the change from array to string intentional?

File: tests/cases/trait/one-per-file [hunk 2]
```diff
@@ -19,4 +19,5 @@ trait Bar
 
 ---MESSAGES---
 11:1 PSR1.Classes.ClassDeclaration.MultipleClasses
+11:1 Squiz.Classes.ClassFileName.NoMatch
```
Comment (stephenwf) [inline tests/cases/trait/one-per-file]: Regression? trait Foo does exist in the file

File: tests/cases/trait/method-defaults (hunk 1 above)
Comment (thewilkybarkid) [reply tests/cases/trait/method-defaults]: Was copied from #32, was just to reduce the amount of changes (as now need to say array of _what_). But that has been done in some places, so will probably tweak to be consistent (not sure which though).

File: tests/cases/trait/one-per-file (hunk 2 above)
Comment (thewilkybarkid) [reply tests/cases/trait/one-per-file]: See #37.

File: src/Libero/Sniffs/Types/IterableTypesSniff.php [hunk 3]
```diff
@@ -0,0 +1,358 @@
...
+     */
+    public function register() : array
+    {
+        return [
+            T_VARIABLE,
+            T_FUNCTION,
+        ];
+    }
+
+    // phpcs:ignore SlevomatCodingStandard.TypeHints.TypeHintDeclaration.MissingParameterTypeHint
```
Comment (thewilkybarkid) [inline src/Libero/Sniffs/Types/IterableTypesSniff.php]: This is annoying. `$pointer` doesn't have a type hint; adding a `@param` here sees it deleted by the fixer. 🤷‍♂️

File: tests/cases/trait/one-per-file (hunk 2 above)
Comment (stephenwf) [reply tests/cases/trait/one-per-file]: Ah cool, makes sense 👍

File: tests/cases/trait/method-defaults (hunk 1 above)
Comment (stephenwf) [reply tests/cases/trait/method-defaults]: cool, so not a bug in the auto-fixer! Makes sense though

File: src/Libero/Sniffs/WhiteSpace/ObjectStaticOperatorSpacingSniff.php [hunk 4]
```diff
@@ -11,7 +11,10 @@
 
 final class ObjectStaticOperatorSpacingSniff extends BaseObjectOperatorSpacingSniff
 {
-    public function register()
+    /**
+     * @return array<int>
```
Comment (thewilkybarkid) [inline src/Libero/Sniffs/WhiteSpace/ObjectStaticOperatorSpacingSniff.php]: Similar to https://github.com/libero/php-coding-standard/pull/40#discussion_r245673354, shouldn't need to re-declare.

**Comment by thewilkybarkid** [comment]
Work is now happening on [slevomat/coding-standard](https://github.com/slevomat/coding-standard) (eg https://github.com/slevomat/coding-standard/commit/4c37991359cfe1b84cf9a46c106f120464de743e) so will change this PR to use that when it's ready.

**Comment by thewilkybarkid** [comment]
[That was quick!](https://github.com/slevomat/coding-standard/releases/tag/5.0.0)

**Comment by thewilkybarkid** [comment]
Not sure about this having tried it out a bit: since files are read in isolation it requires duplicating on an interface and on the implementation, for example. Might be better in [phpstan/phpstan-strict-rules](https://github.com/phpstan/phpstan-strict-rules), but will need to contribute it (refs https://github.com/phpstan/phpstan-strict-rules/pull/35).

**Comment by thewilkybarkid** [comment]
Closing like https://github.com/libero/php-coding-standard/pull/32. Will open a separate PR to require generics-style iterators in doc comments.

**Pull request closed without merging**